src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java

```diff
@@ -620,6 +631,42 @@ public void enforceLabelAttributes()
     }
   }
 
+  @StarlarkMethod(
+      name = "watch_dir",
```

Comment:
fwiw i didn't reply to this thread yesterday because it didn't exist yesterday :P looks like you drafted it but never sent it until today.

but anyway, putting a glob pattern directly in rctx.watch is a bit awkward because you're presumably globbing from the working directory (ie. where the repo is being fetched into). so `rctx.watch(['*.txt'])` wouldn't make sense, but `path.glob(...)` probably would.

in any case i don't think we could realistically piggyback onto GlobFunction as it just works too differently: worrying about package boundaries, returning labels instead of paths, not digesting file contents, etc etc.

regarding folding watch_dir into watch: i ended up going with path.readdir instead, which i think it's a more natural api. we could actually still choose to make rctx.watch on a dir be sensitive to dirents in the future, but i don't think it's as important a question now. let me know if you disagree.

Reply:
My bad it is then. I'm not the most relaxed and well rested I have ever been :)

I have only one concern with the `path.readdir()` idea: we currently have the `read()` / `watch()` dichotomy for files and the "just use `readdir()` for everything" idea makes it so that we don't have it for directories. I'm not sure it matters, though. At the very least, this decision allows you to postpone the decision about whether you want the recursive watching syntax to be the same as that of globbing (I mean, that's probably like a day, but still...) and teaching `readdir()`  to watch the file system is a pretty obvious win.